scripts: attempt to fix some translations in update-translations.py

GChuf opened the pull request:
The script now tries to fix some of the translations which would otherwise be omitted - that is translations of strings containing format specifiers.
Also added a counter for some statistics about translations fixed and languages removed due to minimum translated strings requirement.
In addition, minimum required strings for translations to be accepted bumped from 10 to 100.

GChuf committed:
scripts: attempt to fix some translations in update-translations.py

The script now tries to fix some of the translations which would otherwise be omitted - that is translations of strings containing format specifiers.
Also added a counter for some statistics about translations fixed and languages removed due to minimum translated strings requirement.

GChuf commented:
If code style needs to be changed or anything like that, let me know.
But the script does the job. See the results below:
![Screenshot_62](https://user-images.githubusercontent.com/42591821/63227256-14efd400-c1d4-11e9-8398-f18e913638fe.png)

emilengler commented on contrib/devtools/update-translations.py:
```diff
@@ -169,30 +179,63 @@ def postprocess_translations(reduce_diff_hacks=False):
                     errors = []
                     valid = check_format_specifiers(source, translation, errors, numerus) and not contains_bitcoin_addr(translation, errors)
-
+					
```
Please remove this whitespace

GChuf replied:
thanks, removed

fanquake commented on contrib/devtools/update-translations.py:
```diff
@@ -131,6 +136,11 @@ def contains_bitcoin_addr(text, errors):
+    global tf #translations fixed
+    global lr #languages removed
+    tf = 0
+    lr=0
```
Please be consistent with formatting.

fanquake commented on contrib/devtools/update-translations.py:
```diff
@@ -172,27 +182,60 @@ def postprocess_translations(reduce_diff_hacks=False):
+                           translation_node.text = fix_string(translation_node.text) #fix most common mistakes by replacing symbols
+                           translation_f = split_format_specifiers(find_format_specifiers(translation_node.text))
+                           if source_f == translation_f: # check if translation is acceptable after fixing it. if not, try to fix more.
+                               if translation_node.text[0] != "%" and translation_node.text.find(' %') == -1 and translation_node.text.find('(%') == -1 and translation_node.text.find('\'%') == -1 and translation_node.text.find('\"%') == -1: #if the translation seems okay, add spaces before % if needed - only if certain strings are not found and if '%' is not the first symbol in a string.
```
Instead of adding a very long comment at the end of an already long line of code, you could you bring the it back over above the this line and break it up over multiple lines if needed.

fanquake reviewed (commented):
Left some general comments. Please squash and push a single commit after addressing them all locally.

@laanwj is probably best suited to review this, as he'll know the rationale behind the original `MIN_NUM_MESSAGES` values etc.

I had been thinking this script could be moved to the maintainer tools repo, as it basically only receives random Python cleanups here.

laanwj commented:
Thanks for trying to improve, however

- I don't think `MIN_NUM_MESSAGES` should increased. This threshold is there to ignore useless translations that have only three or so messages. Translating 10 often-seen messages can already be useful. This is even more true for translations that have a language code and a country for a certain dialect.

- The place to make more comprehensive fixes to messages is not here. This amounts to second-guessing translators. Please make the fixes on Transifex itself where possible, so people learn how to translate the messages.

The counters could be useful, I suppose.

> I had been thinking this script could be moved to the maintainer tools repo, as it basically only receives random Python cleanups here.

Agree.

laanwj commented on contrib/devtools/update-translations.py:
```diff
@@ -1,17 +1,18 @@
 # Copyright (c) 2014 Wladimir J. van der Laan
 # Distributed under the MIT software license, see the accompanying
-# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+# file COPYING or http://www.opensource.org/licenses/mit-license.php
```
Please do not make random changes to the license header here. Keep it the same as in every single other file.

GChuf replied:
Sorry for not mentioning my reasoning - working in ubuntu text editor, the link was only displayed properly (italics) when the dot was removed. But I see now other headers all contain dots at the end.
Will add the dot back.

GChuf commented:
@laanwj
About the minimum number of translations, I agree 10 often seen messages can be useful. Whether or not these translations are the ones often seen or not (my thinking was that they're not), it would be best to keep the old requirement of 10 translations anyway.

I didn't understand your second point - what do you mean by comprehensive fixes and second-guessing? The script only fixes strings with wrong format specifiers (but only in the strings that would otherwise be left out by the script itself!). The translation remains the same, all it does is, for example, change _1%_ to _%1_. So out of ~250 translations this script removes, it now fixes ~180 of them, without messing with the actual translations.

However, I did fix some messages on transifex while doing this, but fixing format specifiers is a tedious job and people will continue to "translate" them poorly even after good examples are given.
In arabic language, for example, where they write from right to left, there are a lot of occurrences where the format specifier is inverted. Typos also happen constantly.

And about transifex and translations, I'd like to reference this issue here: https://github.com/bitcoin/bitcoin/issues/16591
There are a lot of duplicate translations, I believe some of them need to be removed. I'd like to help with that.

GChuf commented on contrib/devtools/update-translations.py:
```diff
@@ -29,7 +30,7 @@
 LOCALE_DIR = 'src/qt/locale'
 # Minimum number of messages for translation to be considered at all
-MIN_NUM_MESSAGES = 10
+MIN_NUM_MESSAGES = 100
```
See my comment below

laanwj commented:
> The translation remains the same, all it does is, for example, change 1% to %1. 

I think that's a good aim, but what if someone *means* 1%, say, for a percentage? That's my fear what I mean with second-guessing. I'm often annoyed by tools that try to 'intelligently' guess what I'm trying to do and get it wrong which makes everything so much more complex.

Although you alleviate this concern somewhat by only fixing messages that are detected as broken in the first place. So I'm not against this, just wary. The preferred place to fix messages is upstream, thanks for fixing a few there too :)

GChuf commented:
I understand what you mean now, but yes, I think the overlap between _broken translations with format specifiers_ and _translations using percentages_ is very little (I never found any). In addition, it would have to be exactly 1% or 2%, not x%. Plus, anything that comes out fixed, is a translation that was not included up until now. I think the script works perfectly 99% of the time for the purpose it was built, but I understand the concerns.

I intend to manually fix some or all of the translations in transifex which cant be fixed by this script, although I still don't understand why some translations are NoneType objects - I didnt manage to fix those strings in python.

laanwj commented:
Yes, doing the work manually is time-consuming.

Which gives me an idea, could we batch-fix these (after looking over them once to make sure we're not doing something stupid) through the Transifex API?

GChuf commented:
Squashed and addressed all comments.
I don't know about transifex API, but I agree that it's a good idea if it's possible. Although I would argue patching them in the future with this script would still be useful - unless you were to batch fix all the translations every time, but I imagine this to be done on rare occasions.

GChuf commented:
@laanwj I cleaned up the translations for 0.18 completely. The script now reports 3 mismatches total.

I was also thinking about cleaning some unnecessary locales on transifex (like moving _(it_IT)_ to _(it)_ and removing _(it_IT)_ completely, but I'd need more privileges, of course.

I am also thinking about cleaning 0.17 translations, but only if they get included in the 0.17.2, else it's pointless (also @fanquake).
In the last 2 days some 0.17 translations were updated, but only a few, manually by me. They were not run through the script. So there are some improvements, but I'd like to clean everything if it gets included.

Please let me know what you think.

fanquake commented:
> I am also thinking about cleaning 0.17 translations, but only if they get included in the 0.17.2, else it's pointless

If you update additional translations for 0.17.2 I can pull them into #16638.

fanquake commented:
@GChuf Thanks for your work here, however the script has been moved to the [maintainer tools repo](https://github.com/bitcoin-core/bitcoin-maintainer-tools/blob/master/update-translations.py). 

Can you please re-open your pull request over there?

GChuf commented:
Yes, I'll open a PR request there
